lib/kill.js

```diff
@@ -52,14 +53,33 @@ const spawnedCancel = (spawned, context) => {
 	}
 };
 
+const timeoutKill = (spawned, signal, reject) => {
+	spawned.kill(signal);
+	reject(Object.assign(new Error('Timed out'), {timedOut: true, signal}));
+};
+
 // `timeout` option handling
-const setupTimeout = (spawned, {timeout, killSignal}, context) => {
-	if (timeout > 0) {
-		return setTimeout(() => {
-			context.timedOut = true;
-			spawned.kill(killSignal);
-		}, timeout);
+const setupTimeout = (spawned, {timeout, killSignal = 'SIGTERM'}, spawnedPromise) => {
+	if (timeout === 0 || timeout === undefined) {
+		return spawnedPromise;
+	}
+
+	if (!Number.isInteger(timeout) || timeout < 0) {
+		throw new TypeError(`Expected the \`timeout\` option to be a non-negative integer, got \`${timeout}\` (${typeof timeout})`);
 	}
+
+	let timeoutId;
+	const timeoutPromise = new Promise((resolve, reject) => {
```

Comment:
We unfortunately have to use `new Promise` because `promisify(setTimeout)` does not return the `timeoutId`. However we need that `timeoutId` to clean up the timeout when the child process exits normally.